app/ts-cli/geminicli/export.go

```diff
@@ -0,0 +1,1001 @@
+	fmt.Fprintf(metaWriter, "# FROM TSSP FILE.\n\n")
+	for measurementName, files := range measurementFilesMap {
+		fmt.Fprintf(metaWriter, "# CONTEXT-MEASUREMENT: %s \n", measurementName)
+		for _, file := range files {
+			// ie./tmp/openGemini/data/data/db1/0/autogen/1_1567382400000000000_1567987200000000000_1/tssp/average_temperature_0000/00000002-0000-00000000.tssp
+			splits := strings.Split(file, string(os.PathSeparator))
+			shardDir := splits[len(splits)-4]
+			_, indexId, _, err := engine.ParseShardDir(shardDir)
+			if err != nil {
+				return err
+			}
+			if err := e.writSingleTsspFile(file, outputWriter, indexesMap[indexId]); err != nil {
+				return err
+			}
+		}
+		fmt.Fprintf(outputWriter, "\n")
+	}
+	return nil
+}
+
+// writSingleTsspFile writes a single tssp file's all records.
+func (e *Exporter) writSingleTsspFile(filePath string, outputWriter io.Writer, index *tsi.MergeSetIndex) error {
+	lockPath := ""
+	tsspFile, err := immutable.OpenTSSPFile(filePath, &lockPath, true, false)
+	defer util.MustClose(tsspFile)
+
+	if err != nil {
+		return err
+	}
+	fi := immutable.NewFileIterator(tsspFile, immutable.CLog)
+	itr := immutable.NewChunkIterator(fi)
+
+	for {
+		if !itr.Next() {
+			break
+		}
+		sid := itr.GetSeriesID()
+		if sid == 0 {
+			return fmt.Errorf("series ID is zero")
+		}
+		rec := itr.GetRecord()
+		record.CheckRecord(rec)
+
+		// filter time range
+		if rec.MaxTime(true) < e.filter.startTime || rec.MinTime(true) > e.filter.endTime {
+			continue
+		}
+
+		if err := e.writeSeriesRecords(outputWriter, sid, rec, index); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// writeSeriesRecords writes all records pointed to by one sid.
+func (e *Exporter) writeSeriesRecords(outputWriter io.Writer, sid uint64, rec *record.Record, index *tsi.MergeSetIndex) error {
+
+	var combineKey []byte
+	var seriesKeys [][]byte
+	var isExpectSeries []bool
+	var err error
+	// Use sid get series key's []byte
+	if seriesKeys, _, isExpectSeries, err = index.SearchSeriesWithTagArray(sid, seriesKeys, nil, combineKey, isExpectSeries, nil); err != nil {
+		return err
+	}
+	var series [][]byte
+	for i := range seriesKeys {
+		if !isExpectSeries[i] {
+			continue
+		}
+		bufSeries := influx.GetBytesBuffer()
+		// parse series key to line protocol and escaping special characters to make it valid to import
+		bufSeries, err = Parse2SeriesKeyWithoutVersion(seriesKeys[i], bufSeries, false)
+		if err != nil {
+			return err
+		}
+		series = append(series, bufSeries)
+	}
+
+	var recs []record.Record
+	recs = rec.Split(recs, 1)
+	for _, r := range recs {
+		if err := e.writeSingleRecord(outputWriter, series, r); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// writeSingleRecord parses a record and a series key to line protocol, and writes it.
+func (e *Exporter) writeSingleRecord(outputWriter io.Writer, seriesKey [][]byte, rec record.Record) error {
+	tm := rec.Times()[0]
+	if !e.filter.Filter(tm) {
```

Comment:
Should we judge the time range, not only the minimum time?

Reply:
I filtered the maximum and minimum times in `writeSingleTsspFile()`, but did not encapsulate them as methods.
```go
if rec.MaxTime(true) < e.filter.startTime || rec.MinTime(true) > e.filter.endTime {
	continue
}
``` 
In the latest commit, I encapsulated two new methods